src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializer.java

```diff
@@ -367,6 +367,15 @@ public Object deserializeFromObject(JsonParser p, DeserializationContext ctxt) t
         final Object bean = _valueInstantiator.createUsingDefault(ctxt);
         // [databind#631]: Assign current value, to be accessible by custom deserializers
         p.setCurrentValue(bean);
+
+        // [databind#3838]: Uniform handling of missing objectId
+        if (_objectIdReader != null) {
+            // check if there are any properties present in the JSON object
+            if (!p.hasCurrentToken() || p.getCurrentToken() == JsonToken.END_OBJECT) {
+            // throw an error here since there are no properties in the JSON object
+                ctxt.reportUnresolvedObjectId(_objectIdReader, bean);
+            }
+        }
```

Comment:
Quick note: I think this should probably be moved after following block (`if (p.canReadObjectId()) ...`) since that handles "native" Object Ids (mostly for YAML).